### PR TITLE
RtpsRelay: updated tracking of client network addresses

### PR DESCRIPTION
**Pull request by mitza-oci**

### CODE REVIEW

File: tools/rtpsrelay/GuidAddrSet.cpp [hunk 1]
```diff
@@ -66,31 +66,24 @@ GuidAddrSet::record_activity(const Proxy& proxy,
...
-                   expiration_guid_addr_queue_.size(),
-                   admission_control_queue_.size()));
-      }
-      relay_stats_reporter_.new_address(now);
-      const GuidAddr ga(src_guid, remote_address);
-      expiration_guid_addr_queue_.push_back(std::make_pair(expiration, ga));
+  if (addr_set_stats.upsert_address(remote_address, now, expiration, config_.max_addr_set_size())) {
+    if (config_.log_activity()) {
+      ACE_DEBUG((LM_INFO, "(%P|%t) INFO: GuidAddrSet::record_activity "
+                 "%C %C is at %C %C into session addrs=%B total=%B remote=%B deactivation=%B expire=%B admit=%B\n",
```
Comment (jrw972) [inline tools/rtpsrelay/GuidAddrSet.cpp:72]:
```suggestion
                 "%C %C is at %C %C into session ips=%B total=%B remote=%B deactivation=%B expire=%B admit=%B\n",
```

**Review by jrw972 — CHANGES_REQUESTED** [review]

File: tools/rtpsrelay/GuidAddrSet.cpp [hunk 2]
```diff
@@ -66,31 +66,24 @@ GuidAddrSet::record_activity(const Proxy& proxy,
...
-                   guid_addr_set_map_.size(),
-                   remote_map_.size(),
-                   deactivation_guid_queue_.size(),
-                   expiration_guid_addr_queue_.size(),
-                   admission_control_queue_.size()));
-      }
-      relay_stats_reporter_.new_address(now);
-      const GuidAddr ga(src_guid, remote_address);
-      expiration_guid_addr_queue_.push_back(std::make_pair(expiration, ga));
+  if (addr_set_stats.upsert_address(remote_address, now, expiration, config_.max_addr_set_size())) {
```
Comment (simpsont-oci) [inline tools/rtpsrelay/GuidAddrSet.cpp:69]: Does the config's `max_addr_set_size` also need to be renamed?

File: tools/rtpsrelay/GuidAddrSet.h [hunk 3]
```diff
@@ -10,15 +10,27 @@
...
-typedef std::map<AddrPort, OpenDDS::DCPS::MonotonicTimePoint> AddrSet;
+struct PortSet {
+  std::map<u_short, OpenDDS::DCPS::MonotonicTimePoint> spdp_ports, sedp_ports, data_ports;
+  bool empty() const { return spdp_ports.empty() && sedp_ports.empty() && data_ports.empty(); }
+};
+
+struct InetAddrHash {
+  std::size_t operator()(const ACE_INET_Addr& addr) const noexcept
+  {
+    return addr.hash();
```
Comment (simpsont-oci) [inline tools/rtpsrelay/GuidAddrSet.h:25]: Just a heads up... this hashing is not great, especially for ipv6.

File: tools/rtpsrelay/GuidAddrSet.cpp (hunk 2 above)
Comment (mitza-oci) [reply tools/rtpsrelay/GuidAddrSet.cpp:69]: Probably

File: tools/rtpsrelay/GuidAddrSet.h (hunk 3 above)
Comment (mitza-oci) [reply tools/rtpsrelay/GuidAddrSet.h:25]: I'm basing it on `typedef std::unordered_map<Remote, OpenDDS::DCPS::GUID_t, RemoteHash> RemoteMap` in the existing code.  Maybe they could both be improved in a future PR.

**Review by jrw972 — APPROVED** [review]

**Pull request merged**